src/state_machine.py

```diff
@@ -0,0 +1,98 @@
+#!/usr/bin/env python3
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Encapsulation of the charm's state machine."""
+
+import codecs
+import pickle
+
+from ops import Object
+
+import transitions
+
+
+class StateMachine:
+    states = [
+        "waiting_for_database",
+        "ready",
+        "writing",
+    ]
+
+    state_transitions = [
+        ("waiting_for_database", "ready"),
+        ("ready", "writing"),
+        ("writing", "ready"),
+    ]
+
+    blocking_states = [
+        "waiting_for_database",
+    ]
+
+    def __init__(self):
+        self.machine = transitions.Machine(
+            model=self,
+            states=self.states,
+            initial="waiting_for_database",
+        )
+
+        for (source, destination) in self.state_transitions:
+            self.machine.add_transition(trigger=f"to_{destination}", source=source, dest=destination)
+
+
+class CharmState(Object):
+    def __init__(self, charm):
+        super().__init__(charm, "state-machine")
+
+        self.charm = charm
+        if not self.charm.unit_peer_data:
+            return
+
+        if self.charm.unit_peer_data.get("state-machine"):
+            self.machine = pickle.loads(
+                codecs.decode(self.charm.unit_peer_data.get("state-machine").encode(), "base64")
+            )
+        else:
+            self.machine = StateMachine()
+
+        self.framework.observe(
+            self.charm.on.get_state_action, self._get_state_action
+        )
+
+    def __del__(self) -> None:
+        """Store the state machine in the databag."""
+        if not self.charm.unit_peer_data:
+            return
+
+        self.charm.unit_peer_data["state-machine"] = codecs.encode(
+            pickle.dumps(self.machine), "base64"
+        ).decode()
+
+    @property
+    def is_blocked(self) -> bool:
+        """Indicate whether the state is blocked."""
+        return self.machine.state in ["waiting_for_database"]
+
+    @property
+    def current_state(self) -> str:
```

Comment:
I would expect the current state to have some assertion step, meaning that, for our use case, the state can change independently from the state machine existing transitions and we need to validate one against the other.
In a way, this can be the trigger for a reconciliation step, when states differ.
E.g.: `current_state` is "writing" but scripts has crashed. A reconciliation would evaluate it "wrinting" state pre-conditions are met and restart the script.